### PR TITLE
himalaya: adjust module for `v1.0.0-beta`

### PR DESCRIPTION
**Pull request by soywod**
### Description

Adjust module to match the `v1.0.0-beta`. I just open the PR as draft (waiting for https://github.com/NixOS/nixpkgs/pull/278088).

### Checklist

- [ ] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [ ] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

#### Maintainer CC

@toastal

### CODE REVIEW

File: modules/programs/himalaya.nix [hunk 1]
```diff
@@ -8,24 +8,52 @@ let
...
       imapEnabled = !isNull account.imap && !notmuchEnabled;
-      maildirEnabled = !isNull account.maildir && !imapEnabled
-        && !notmuchEnabled;
+      maildirEnabled = !isNull account.maildir && !notmuchEnabled
+        && !imapEnabled;
+      smtpEnabled = !isNull account.smtp;
+      sendmailEnabled = !smtpEnabled && !isNull account.msmtp;
+
+      backendConfig = if notmuchEnabled then {
+        backend = "notmuch";
```
[the remark this answers is not in the document]
Comment (soywod) [reply modules/programs/himalaya.nix:37]: Is there any way to do pattern matching in Nix? Like `switch` in JavaScript or `match` in Rust?

**Comment by soywod** [comment]
> I didn’t expect watch to be added already

The watch (and notify) were already there before, I just refactored the code to match the new watch API which combines both previous watch (running shell commands) and previous notify (send system notifications).

I agree about the `if`, I will propose an alternative.

Thank you for your feedback!

**Comment by soywod** [comment]
> @soywod do you have a recent rebase from the default branch atop this?

I just did it.

**Comment by erooke** [comment]
For whatever it's worth while trying to rebuild my configuration off of this branch I got the following error
```
error:
       … while calling the 'derivationStrict' builtin

         at /builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'home-manager-generation'
         whose name attribute is located at /nix/store/9s5qs4hni9fj88x79iw6im7amv7ghb76-source/pkgs/stdenv/generic/make-derivation.nix:352:7

       … while evaluating attribute 'buildCommand' of derivation 'home-manager-generation'

         at /nix/store/9s5qs4hni9fj88x79iw6im7amv7ghb76-source/pkgs/build-support/trivial-builders/default.nix:98:16:

           97|         enableParallelBuilding = true;
           98|         inherit buildCommand name;
             |                ^
           99|         passAsFile = [ "buildCommand" ]

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: attribute 'accounts' missing

       at /nix/store/kgyb4snyic74z8jyybx1zclh9dk70q8n-source/modules/programs/himalaya.nix:68:33:

           67|         backend = "notmuch";
           68|         notmuch.database-path = config.accounts.email.maildirBasePath;
             |                                 ^
           69|       });
```
Here is a minimal config which seems to trigger the same breakage:
```nix
{...}: {
  home = {
    stateVersion = "24.05";
    username = "test";
    homeDirectory = /home/test;
  };

  accounts.email = {
    maildirBasePath = ".mail";

    accounts.test = {
      primary = true;
      address = "test@mail.com";
      userName = "test@mail.com";
      realName = "Test Testerton";

      notmuch.enable = true;
      himalaya.enable = true;
    };
  };
  programs.notmuch.enable = true;
  programs.himalaya.enable = true;
}
```
This was using the current `nixpkgs-unstable` and `github:soywod/home-manager/himalaya-v1.0.0-beta`. If needed I could supply a `flake.lock` and `flake.nix` but that felt like un-needed noise.

**Comment by soywod** [comment]
> For whatever it's worth while trying to rebuild my configuration off
> of this branch I got the following error

I pushed a fix, it should be good now!

**Comment by erooke** [comment]
> I pushed a fix, it should be good now!

Yup the home manager config now builds on my end! Himalaya seems unhappy with the generated configuration now though.

```
$ himalaya accounts list

Error: cannot parse config file at "/home/ethan/.config/himalaya/config.toml"

Caused by:
    TOML parse error at line 1, column 1
      |
    1 | [rooke]
      | ^
    unknown variant `notmuch`, expected one of `maildir`, `imap`, `smtp`, `sendmail`    
```

**Comment by soywod** [comment]
> unknown variant `notmuch`, expected one of `maildir`, `imap`, `smtp`,
> `sendmail`

The Notmuch backend is not available by default, it requires the
"notmuch" cargo feature.

The problem is that the Notmuch backend was temporarily be disabled in
the "v1.0.0-beta" (which is the one available on nixpkgs-unstable). It
has been reactivated in the "v1.0.0-beta.2", which is not yet available
on nixpkgs-unstable:

<https://github.com/NixOS/nixpkgs/pull/284358>

Once the "v1.0.0-beta.2" lands on nixpkgs-unstable, you will be able to
adjust your configuration this way:

programs.himalaya = {
  enable = true;
  package = pkgs.himalaya.override { buildFeatures = ["notmuch"]; };
};

**Comment by soywod** [comment]
Does it make sense to force the "notmuch" cargo feature from the
home-manager module? Something like:

home.packages = [ himalaya.package.override {
  buildFeatures = lib.optional notmuchEnabled "notmuch";
} ]

**Comment by erooke** [comment]
I personally would like that, maybe just complaining if it's not enabled would also work. I don't know how other hm modules handle it or if it makes sense in the broader picture I'm afraid.

Another possibility could be complaining from Himalaya that it was compiled without notmuch support. I did not find the current error very enlightening.

Jan 28, 2024 2:39:12 PM Clément DOUIN ***@***.***>:

**Comment by soywod** [comment]
@rycee what is the best approach for this kind of situation?

**Comment by soywod** [comment]
> I got the config to build but …

There is no way to know the correct location of errors at the moment. It
is a known problem, see:

- <https://github.com/toml-rs/toml/issues/589#issuecomment-1872345017>
- <https://github.com/serde-rs/serde/issues/1183>

Could you share your home-manager config as well as the generated TOML?

**Comment by soywod** [comment]
> Well that kinda sucks ha.

Yes, I know. It is really frustrating because the issue does not come
from Himalaya. But I am working on a work around, I let you know!

**Comment by soywod** [comment]
I cannot understand why `.override` does not work from the home manager module but work outside of it: ` error: attribute 'override' missing` https://github.com/nix-community/home-manager/actions/runs/7702476916/job/20990851373?pr=4839

**Comment by rycee** [comment]
All good to merge?

**Comment by rycee** [comment]
Cool, thanks! Merged to master now 🙂

**Pull request merged**